File: src/WAS/Common/Constants/Permissions.cs
using WAS.Attributes;

namespace WAS.Common.Constants
{
    public static class Permissions
    {
        // 기준 정보
        [PermissionInfo("기준 정보 조회")]
        public const string MasterDataView = "Permissions.MasterData.View";
        [PermissionInfo("기준 정보 관리/수정")]
        public const string MasterDataEdit = "Permissions.MasterData.Edit";

        // 재고 관리
        [PermissionInfo("재고 현황 조회")]
        public const string InventoryView = "Permissions.Inventory.View";
        [PermissionInfo("재고 관리/출하")]
        public const string InventoryEdit = "Permissions.Inventory.Edit";

        // 공정 관리
        [PermissionInfo("공정/작업지시 조회")]
        public const string ProcessView = "Permissions.Process.View";
        [PermissionInfo("작업지시 발행/공정 수행")]
        public const string ProcessExecute = "Permissions.Process.Execute";

        // 품질 관리
        [PermissionInfo("품질 검사 조회")]
        public const string QCView = "Permissions.QC.View";
        [PermissionInfo("품질 검사 판정/수행")]
        public const string QCExecute = "Permissions.QC.Execute";
    }
}

File: src/Frontend/src/store/permissions.ts
import type { UserInfo } from '../api/data-contracts';

export const PERMISSIONS = {
  masterDataView: 'Permissions.MasterData.View',
  masterDataEdit: 'Permissions.MasterData.Edit',
  inventoryView: 'Permissions.Inventory.View',
  inventoryEdit: 'Permissions.Inventory.Edit',
  processView: 'Permissions.Process.View',
  processExecute: 'Permissions.Process.Execute',
  qcView: 'Permissions.QC.View',
  qcExecute: 'Permissions.QC.Execute',
} as const;

interface AccessRule {
  allowedPermissions?: string[];
  allowedRoles?: string[];
}

const ROLE_PERMISSION_FALLBACKS: Record<string, string[]> = {
  admin: ['*'],
  operator: [PERMISSIONS.processView, PERMISSIONS.processExecute],
  qc: [PERMISSIONS.qcView, PERMISSIONS.qcExecute],
  viewer: [PERMISSIONS.processView, PERMISSIONS.qcView, PERMISSIONS.masterDataView, PERMISSIONS.inventoryView],
};

const normalize = (value?: string | null) => value?.trim().toLowerCase() || '';

export const hasAccess = (user: UserInfo | null, rule: AccessRule = {}) => {
  const userWithLegacyKeys = user as (UserInfo & {
    RoleCode?: string | null;
    Permissions?: string[] | null;
  }) | null;
  const roleCode = normalize(user?.roleCode || userWithLegacyKeys?.RoleCode || 'VIEWER');
  const userPermissions = user?.permissions || userWithLegacyKeys?.Permissions || [];
  const permissions = (userPermissions.length > 0
    ? userPermissions
    : ROLE_PERMISSION_FALLBACKS[roleCode] || []
  ).map(normalize);
  const allowedPermissions = rule.allowedPermissions?.map(normalize) || [];
  const allowedRoles = rule.allowedRoles?.map(normalize) || [];

  if (roleCode === 'admin' || permissions.includes('*')) {
    return true;
  }

  if (allowedPermissions.length > 0) {
    return allowedPermissions.some((permission) => permissions.includes(permission));
  }

  if (allowedRoles.length > 0) {
    return allowedRoles.includes(roleCode);
  }

  return true;
};
